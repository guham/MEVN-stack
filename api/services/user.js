const { User } = require('../models');

exports.findOrCreate = async (userSub) => {
  let user = await User.findOne({ sub: userSub }).exec();
  if (!user) {
    user = new User({ sub: userSub });
    await user.save();
  }
  return user;
};

exports.storeUserRefreshToken = async (userSub, refreshToken) => {
  const user = await exports.findOrCreate(userSub);
  const timestamp = new Date().getTime();
  user.refreshTokens.set(timestamp.toString(), refreshToken);
  await user.save();
  return user;
};

/**
 * @param {object} user
 * @param {string} refreshToken
 * @returns {boolean} true if the refresh token existed and has been removed, otherwise false
 */
exports.removeUserRefreshToken = async (user, refreshToken) => {
  const tokens = user.refreshTokens;
  const entry = Array.from(tokens).filter(element => element[1] === refreshToken).shift();
  if (!entry) {
    return false;
  }

  tokens.delete(entry[0]);
  await user.save();
  return true;
};
