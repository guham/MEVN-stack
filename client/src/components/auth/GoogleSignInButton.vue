<template>
  <button
    v-if="isAuthenticated"
    id="btn-signout"
    class="btn btn-grey"
    type="button"
    @click="signOutAndRedirect()">
    Sign Out
  </button>
  <button
    v-else
    id="btn-signin"
    class="btn btn-green"
    type="button"
    @click="signIn()">
    Sign In
  </button>
</template>

<script>
import { createNamespacedHelpers } from 'vuex';

const { mapGetters, mapActions } = createNamespacedHelpers('user');

export default {
  name: 'GoogleSignInButton',

  computed: {
    ...mapGetters([
      'isAuthenticated',
    ]),
  },

  methods: {
    ...mapActions([
      'signIn',
      'signOut',
    ]),
    async signOutAndRedirect() {
      await this.signOut();
      this.$router.replace({ name: 'Home' });
    },
  },
};
</script>
