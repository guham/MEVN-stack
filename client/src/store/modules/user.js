import authClient from '@/api/auth';
import * as types from '@/store/mutation-types';
import Notification from '@/models/notification';

const state = {
  isAuthenticated: false,
  userIsSigningIn: false,
  jwt: localStorage.getItem('token'),
  jwtExpiration: parseInt(localStorage.getItem('tokenExpiration'), 10) || null,
};

const getters = {
  isAuthenticated: state => state.isAuthenticated,
  userIsSigningIn: state => state.userIsSigningIn,
  token: state => state.jwt,
};

const mutations = {
  [types.SET_USER_AUTHENTICATED](state) {
    state.isAuthenticated = true;
  },

  [types.SET_USER_IS_SIGNING_IN](state, flag = true) {
    state.userIsSigningIn = flag;
  },

  [types.SET_USER_UNAUTHENTICATED](state) {
    state.isAuthenticated = false;
    state.jwt = null;
    state.jwtExpiration = null;
    localStorage.removeItem('token');
    localStorage.removeItem('tokenExpiration');
  },

  [types.SET_JWT](state, token) {
    state.jwt = token;
    localStorage.setItem('token', token);
  },

  [types.SET_JWT_EXPIRATION](state, tokenExpiration) {
    state.jwtExpiration = tokenExpiration;
    localStorage.setItem('tokenExpiration', tokenExpiration);
  },
};

const actions = {
  async signIn({ dispatch, commit }) {
    commit(types.SET_USER_IS_SIGNING_IN, true);

    try {
      const auth2 = window.gapi.auth2.getAuthInstance();
      const googleUser = await auth2.signIn();
      const idToken = googleUser.getAuthResponse().id_token;
      const response = await authClient.sendIdToken(idToken);
      commit(types.SET_USER_AUTHENTICATED);
      commit(types.SET_JWT, response.token);
      commit(types.SET_JWT_EXPIRATION, response.tokenExpiration);
    } catch (e) {
      // display a notification only when error happened with the API
      // auth2.signIn() will returns an object containing an *error* property if an error happened
      if (!e.error) {
        dispatch('addThenRemoveNotification', new Notification('error', '%SOMETHING_WENT_WRONG%'), { root: true });
      }
      await dispatch('signOut');
    }
    commit(types.SET_USER_IS_SIGNING_IN, false);
  },
  async signOut({ dispatch, commit }) {
    const auth2 = window.gapi.auth2.getAuthInstance();
    await auth2.signOut().catch(() => {
      dispatch('addThenRemoveNotification', new Notification('error', '%SOMETHING_WENT_WRONG%'), { root: true });
    });
    commit(types.SET_USER_UNAUTHENTICATED);
  },
};

export default {
  namespaced: true,
  state,
  getters,
  mutations,
  actions,
};
