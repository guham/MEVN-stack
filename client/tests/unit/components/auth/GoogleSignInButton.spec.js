import Vuex from 'vuex';
import flushPromises from 'flush-promises';
import GoogleSignInButton from '@/components/auth/GoogleSignInButton.vue';
import user from '@/store/modules/user';
import factory from '../../factory';

describe('GoogleSignInButton.vue', () => {
  let store;
  let actions;

  beforeEach(() => {
    actions = {
      signIn: jest.fn(),
      signOut: jest.fn(),
    };

    store = new Vuex.Store({
      modules: {
        user: {
          namespaced: user.namespaced,
          state: user.state,
          getters: user.getters,
          mutations: user.mutations,
          actions,
        },
      },
    });
  });

  test('should display Sign In button when user in not logged', () => {
    const wrapper = factory(GoogleSignInButton, store);
    expect(wrapper.vm.isAuthenticated).toBeFalsy();
    expect(wrapper.find('#btn-signin').isVisible()).toBeTruthy();
    expect(wrapper.find('#btn-signout').length).toBeFalsy();
  });

  test('should display Sign Out button when user is logged', () => {
    store = new Vuex.Store({
      modules: {
        user: {
          namespaced: user.namespaced,
          state: user.state,
          getters: {
            isAuthenticated: () => true,
          },
          mutations: user.mutations,
          actions,
        },
      },
    });

    const wrapper = factory(GoogleSignInButton, store);
    expect(wrapper.vm.isAuthenticated).toBeTruthy();
    expect(wrapper.find('#btn-signout').isVisible()).toBeTruthy();
    expect(wrapper.find('#btn-signin').length).toBeFalsy();
  });

  test('calls store action `signIn` when Sign In button is clicked', () => {
    const wrapper = factory(GoogleSignInButton, store);
    wrapper.find('#btn-signin').trigger('click');
    expect(actions.signIn).toHaveBeenCalled();
  });

  test('emit `signIn` event when Sign In button is clicked', async () => {
    const wrapper = factory(GoogleSignInButton, store);
    wrapper.find('#btn-signin').trigger('click');
    await flushPromises();
    expect(wrapper.emitted('signIn')).toBeTruthy();
    expect(wrapper.emitted('signIn').length).toBe(1);
    expect(wrapper.emitted('signIn')[0]).toEqual([]);
  });

  test('calls store action `signOut` when Sign Out button is clicked', async () => {
    store = new Vuex.Store({
      modules: {
        user: {
          namespaced: user.namespaced,
          state: user.state,
          getters: {
            isAuthenticated: () => true,
          },
          mutations: user.mutations,
          actions,
        },
      },
    });

    const wrapper = factory(GoogleSignInButton, store);
    const spySignOutAndRedirect = jest.spyOn(wrapper.vm, 'signOutAndRedirect');
    const spyRouterReplace = jest.spyOn(wrapper.vm.$router, 'replace');

    wrapper.find('#btn-signout').trigger('click');
    await flushPromises();
    expect(spySignOutAndRedirect).toHaveBeenCalled();
    expect(actions.signOut).toHaveBeenCalled();
    expect(spyRouterReplace).toHaveBeenCalledWith({ name: 'Home', params: {}, path: '' });
  });

  test('has the expected html structure', () => {
    const wrapper = factory(GoogleSignInButton, store);
    const template = wrapper.html();
    expect(template).toMatchSnapshot();
  });
});
