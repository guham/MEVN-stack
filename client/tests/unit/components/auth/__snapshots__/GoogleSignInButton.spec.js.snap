// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleSignInButton.vue has the expected html structure 1`] = `
<button id="btn-signin" type="button" class="btn btn-green">
  Sign In
</button>
`;
